### PR TITLE
[Scala3] Don't show a tooltip on hover if cursor is not on a symbol

### PR DESCRIPTION
**Pull request by tanishiking**
Partially fix https://github.com/scalameta/metals/issues/1707 (hover part of the issue), taking over from https://github.com/scalameta/metals/pull/3743

FYI @tdudzik I opened a PR to fix the issue on the Hover part, if you're interested in working on go-to-definition part, please let me know :) 

This PR fixes Scala3 HoverProvider not to show a tooltip on hover if the cursor is not on a symbol by checking the `namePos` (for example, `namePos` of `def foo(x: Int) = { ... }` is a position of `def <foo>(...) = { ... }` part, not a entire `<def foo(x: Int) = { ... }>`) of the enclosing tree actually contains the cursor position.

Calculate the `namePos` using [SourceTree#namePos](https://github.com/lampepfl/dotty/blob/d8e50f249ce0206faae35d5d3d28f45f1e0f58d1/compiler/src/dotty/tools/dotc/interactive/SourceTree.scala#L21-L22).

no-hover-on-non-symbol

### CODE REVIEW

**Review by dos65 — APPROVED** [review]
LGTM!

File: mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala [hunk 1]
```diff
@@ -49,8 +51,45 @@ object HoverProvider:
...
+        .headOption
+        .getOrElse(EmptyTree)
+      contains(enclosing)
+    end isHoveringOnName
+
+    if tp.isError || tpw == NoType || tpw.isError || path.isEmpty ||
+      (pos.start == pos.end && !isHoveringOnName(
+        enclosing,
+        pos
+      )) // don't check isHoveringOnName for RangeHover
```
Comment (kpodsiad) [inline mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala:91]: Could you extract this part of the condition to `def`?

**Review by kpodsiad — APPROVED** [review]

File: mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala (hunk 1 above)
Comment (tanishiking) [reply mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala]: Refactored a bit in https://github.com/scalameta/metals/pull/3792/commits/cf020525fb33f675ac665bd090553bdee7db99c4

File: mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala [hunk 2]
```diff
@@ -49,8 +51,43 @@ object HoverProvider:
...
+     * ```
+     * @param path - path to the position given by `Interactive.pathTo`
+     */
+    def isHoveringOnName(path: List[Tree], sourcePos: SourcePosition): Boolean =
+      def contains(tree: Tree): Boolean = tree match
+        case tree: NameTree =>
+          source.fold(false) { s =>
+            SourceTree(tree, s).namePos.contains(sourcePos)
+          }
+        // TODO: check the positions for NamedArg and Import
```
Comment (tgodzik) [inline mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala:73]: @tanishiking should it have been addressed in this PR before merging?

File: tests/cross/src/test/scala/tests/hover/HoverTermSuite.scala [hunk 3]
```diff
@@ -112,7 +112,7 @@ class HoverTermSuite extends BaseHoverSuite {
   )
 
   check(
-    "interpolator-apply",
+    "interpolator-apply".tag(IgnoreScala3),
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/hover/HoverTermSuite.scala:115]: This seems to have worked before, why was it ignored?

File: mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala (hunk 2 above)
Comment (tanishiking) [reply mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala:73]: The drawback of not checking these trees is: we show a hover tooltip if the cursor is on

- ` = ` part of named argument
- ` as ` part of import statement

which aren't big problems, so I believe we can address these in another PR (maybe we should open another issue)

Screen Shot 2022-04-08 at 12 39 08

Screen Shot 2022-04-08 at 12 39 00

File: tests/cross/src/test/scala/tests/hover/HoverTermSuite.scala (hunk 3 above)
Comment (tanishiking) [reply tests/cross/src/test/scala/tests/hover/HoverTermSuite.scala:115]: Oh, I guess I overlooked it. It should be fixed!

File: mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala (hunk 2 above)
Comment (tgodzik) [reply mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala:73]: For sure this is not a big issue :+1:

**Pull request merged**